### PR TITLE
pytest migration

### PR DESCRIPTION
**Pull request by shimizukawa**
### Feature or Bugfix
- Feature

### Purpose
- easier to maintain
- using modern test runner
- pytest is active project

### Detail
- replace nose with pytest

### Next stage (not this PR)
- replace deprecations in util.py with pytest features (gen_with_app, yield testing, assert, and more)
- refactoring tests
- test speedup

### CODE REVIEW

File: tests/test_build_html.py
Comment (shimizukawa) [inline tests/test_build_html.py:29]: I think this warning line will never appear because autodoc.rst is not exist in the `roots/test-warnings`.

File: tests/test_build_html.py
Comment (shimizukawa) [inline tests/test_build_html.py:37]: emitted warning with using pytest have `\r` char.

File: tests/test_websupport.py
Comment (shimizukawa) [inline tests/test_websupport.py:128]: re-ordering the test execution as `nose` did.
I'll rewrite such 'ordering depended test cases' at next REFACTORING PR.

File: tests/test_build.py
Comment (shimizukawa) [inline tests/test_build.py:79]: now `tmpdir` is naked pytest fixture, otherwise `tempdir` is temporary directory object that is wrapped with `path`

File: tests/conftest.py
Comment (tk0miya) [inline tests/conftest.py]: Why not `app`? Is there special meaning for the underscore?

File: .travis.yml
Comment (shimizukawa) [inline .travis.yml:15]: `--durations 25` option will indicate like bellow:
```
============================================================================================================================================ slowest 25 test durations =============================================================================================================================================
19.39s call     tests/test_build.py::test_build_all::[16]
6.93s call     tests/test_setup_command.py::test_build_sphinx_return_nonzero_status
6.77s call     tests/test_build_linkcheck.py::test_defaults
6.49s call     tests/test_build_linkcheck.py::test_anchors_ignored
6.06s call     tests/test_websupport.py::test_build
5.76s call     tests/test_build.py::test_build_all::[0]
3.88s call     tests/test_build_html.py::test_html_sourcelink_suffix
2.81s call     tests/test_build.py::test_build_all::[11]
2.55s call     tests/test_build.py::test_build_all::[10]
2.21s call     tests/test_build.py::test_build_all::[1]
2.08s call     tests/test_build.py::test_build_all::[5]
2.02s call     tests/test_build.py::test_build_all::[7]
2.01s call     tests/test_setup_command.py::test_build_sphinx_warning_return_zero_status
1.99s call     tests/test_build.py::test_build_all::[9]
1.99s call     tests/test_build.py::test_build_all::[8]
1.91s call     tests/test_build.py::test_build_all::[4]
1.89s call     tests/test_websupport.py::test_no_srcdir
1.84s call     tests/test_setup_command.py::test_build_sphinx_with_nonascii_path
1.83s call     tests/test_setup_command.py::test_build_sphinx
1.74s call     tests/test_searchadapters.py::test_whoosh
1.72s call     tests/test_setup_command.py::test_build_sphinx_warning_is_error_return_nonzero_status
1.72s call     tests/test_build_latex.py::test_writer
1.47s call     tests/test_build_latex.py::test_footnote
1.42s call     tests/test_build.py::test_build_all::[2]
1.41s call     tests/test_build_texinfo.py::test_texinfo
========================================================================================================================== 980 passed, 11 skipped, 654 pytest-warnings in 209.86 seconds ===========================================================================================================================
```

File: tests/test_metadata.py
Comment (shimizukawa) [inline tests/test_metadata.py:54]: When test was be failure, pytest will show the error as:
```
===================================================================================================================================================== FAILURES =====================================================================================================================================================
___________________________________________________________________________________________________________________________________________________ test_docinfo ___________________________________________________________________________________________________________________________________________________

app = <SphinxTestApp buildername='pseudoxml'>, status = <_io.StringIO object at 0x7f90a0ea2d38>, warning = <_io.StringIO object at 0x7f90a0e45f78>

    @with_app('pseudoxml')
    def test_docinfo(app, status, warning):
        """
        Inspect the 'docinfo' metadata stored in the first node of the document.
        Note this doesn't give us access to data stored in subsequence blocks
        that might be considered document metadata, such as 'abstract' or
        'dedication' blocks, or the 'meta' role. Doing otherwise is probably more
        messing with the internals of sphinx than this rare use case merits.
        """
        app.builder.build(['metadata'])
        env = app.env
        exampledocinfo = env.metadata['metadata']
        expecteddocinfo = {
            'author': u'David Goodger',
            'authors': [u'Me', u'MYself', u'I'],
            'address': u'123 Example Street\nExample, EX  Canada\nA1B 2C3',
            'field name': u'This is a generic bibliographic field.',
            'field name 2': (u'Generic bibliographic fields may contain multiple '
                             u'body elements.\n\nLike this.'),
            'status': u'This is a "work in progress"',
            'version': u'1',
            'copyright': (u'This document has been placed in the public domain. '
                          u'You\nmay do with it as you wish. You may copy, modify,'
                          u'\nredistribute, reattribute, sell, buy, rent, lease,\n'
                          u'destroy, or improve it, quote it at length, excerpt,\n'
                          u'incorporate, collate, fold, staple, or mutilate it, or '
                          u'do\nanything else to it that your or anyone else\'s '
                          u'heart\ndesires.'),
            'contact': u'goodger@python.org',
            'date': u'2006-05-21',
            'organization': u'humankind',
            'revision': u'4564',
            'tocdepth': 1,
            'orphan': u'',
            'nocomments': u'',
        }
>       assert exampledocinfo == expecteddocinfo
E       assert {'address': '...hon.org', ...} == {'address': '1...hon.org', ...}
E         Omitting 14 identical items, use -v to show
E         Differing items:
E         {'authors': ['Me', 'Myself', 'I']} != {'authors': ['Me', 'MYself', 'I']}
E         Use -v to get the full diff

tests/test_metadata.py:54: AssertionError
============================================================================================================================================= 1 failed in 3.49 seconds =============================================================================================================================================
```

File: tests/conftest.py
Comment (shimizukawa) [reply tests/conftest.py]: The name of `app` is used as a fixture at module level.
So I want to avoid overriding the name.

Comment (tk0miya) [reply tests/conftest.py]: Oh, fixture affect whole of module? I don't know that. I understand.

File: tests/conftest.py
Comment (tk0miya) [inline tests/conftest.py]: It seems `testenv` is not used. Is this needed?

File: tests/path.py
Comment (tk0miya) [inline tests/path.py]: AFAIK, `repr_as` is introduced to suppress contents on 1) list of testcases and 2) failure report.

 1) is like this:
https://travis-ci.org/sphinx-doc/sphinx/jobs/188727836#L517


I agree the second purpose make hard to debug on failure.
But this change messes up the merit of `repr_as`.

If pytest shows human-readable list of testcases, `repr_as` is no longer needed. So we should remove it.

File: tests/conftest.py
Comment (tk0miya) [reply tests/conftest.py]: Ah, sorry. I'd misread. You said about the name confliction between global fixture, right?
I understand that.
Okay, either's fine :-)

File: tests/conftest.py
Comment (shimizukawa) [reply tests/conftest.py]: Maybe it is not needed for now. OK, I'll remove this.

File: tests/test_setup_command.py
Comment (tk0miya) [inline tests/test_setup_command.py]: It seems `nonascii_srcdir` is not used. Is it proper as an argument? How about `usefixture`?

File: tests/test_build.py
Comment (tk0miya) [inline tests/test_build.py]: Is this needed?

File: tests/test_build_gettext.py
Comment (tk0miya) [inline tests/test_build_gettext.py]: This should be replaced by `pytest.mark.sphinx` because you marked `with_app` decorator as deprecated.

File: tests/path.py
Comment (shimizukawa) [reply tests/path.py]: Confirmed.
1) is not appeared with pytest: https://travis-ci.org/sphinx-doc/sphinx/jobs/188570306#L1829
2) is bad side-effect of suppressing (1).

> If pytest shows human-readable list of testcases, repr_as is no longer needed. So we should remove it.

OK. I'll remove it.

File: tests/test_build.py
Comment (shimizukawa) [reply tests/test_build.py]: I just left it to minimize the differences, but let's delete it as well.

File: tests/conftest.py
Comment (tk0miya) [inline tests/conftest.py]: Why yield-ing?

Comment (tk0miya) [reply tests/conftest.py]: If you need to use yield here, please add comment.
(for example, "use yield to clean up application object after testing")

File: tests/conftest.py
Comment (tk0miya) [inline tests/conftest.py:77]: Are these clean-up process executed certainly by pytest?
I usually uses try-finally clause to do that.

File: tests/test_build_html.py
Comment (tk0miya) [inline tests/test_build_html.py:663]: Why are you still using `@gen_with_app`? It seems marked as deprecated.
Do you have plan to migrate them?

File: tests/test_build_latex.py
Comment (tk0miya) [inline tests/test_build_latex.py:93]: Note: we have to convert this to fixture.

File: tests/util.py
Comment (tk0miya) [inline tests/util.py:349]: I think `test/util.py` is not public library. So we don't need to keep compatibility.
Is this really needed?

**Review by tk0miya — APPROVED** [review]
LGTM with nits

File: tests/conftest.py
Comment (shimizukawa) [reply tests/conftest.py:77]: In current (nose) testing, some sphinx project for testing insert path on conf.py. I think developers can't do cleanup correctly for such case because it is difficult to notice.

File: tests/conftest.py
Comment (shimizukawa) [reply tests/conftest.py]: There is no special meaning. yield and return is same meaning. (and next PR I will put teardown process here). Anyway, I will change this line into `return` for this PR.

File: tests/test_build_html.py
Comment (shimizukawa) [reply tests/test_build_html.py:663]: If I remove gen_with_app, this PR became as same as #3300.
It seems that you have fully understood the pytest code, so I think it's time to review #3300.

File: tests/util.py
Comment (shimizukawa) [reply tests/util.py:349]: I think it's needed for non-merged-PRs, for a while.

**Comment by tk0miya** [comment]
Note:
This PR is too large to review on updating.
To migrate gradually, I'd like to merge this once and modify the tests step by step.

File: tests/test_build_latex.py
Comment (shimizukawa) [reply tests/test_build_latex.py:93]: Next PR, I will (#3300 trial PR did it)

File: tests/conftest.py
Comment (shimizukawa) [reply tests/conftest.py:77]: Sorry, I misread it.
After yield lines will be executed even if test function raise exception. So this implementation is OK.

**Comment by shimizukawa** [comment]
OK, I'll merge this later today.

File: tests/test_build_html.py
Comment (tk0miya) [reply tests/test_build_html.py:663]: Okay, go ahead step by step.
I will review it after merging this.

File: tests/util.py
Comment (tk0miya) [reply tests/util.py:349]: i understand.

File: tests/conftest.py
Comment (tk0miya) [reply tests/conftest.py:77]: I see

File: tests/conftest.py
Comment (tk0miya) [reply tests/conftest.py]: +1 > return

**Comment by shimizukawa** [comment]
I'll merge this PR into stable after test passing.

**Pull request merged**
